src/expr/impl/src/scalar/string.rs

```diff
@@ -472,6 +474,139 @@ pub fn right(s: &str, n: i32, writer: &mut impl Write) {
         .for_each(|c| writer.write_char(c).unwrap());
 }
 
+/// `quote_literal(string text)`
+///
+/// Return the given string suitably quoted to be used as a string literal in an SQL statement
+/// string. Embedded single-quotes and backslashes are properly doubled.
+/// Note that `quote_literal` returns null on null input; if the argument might be null,
+/// `quote_nullable` is often more suitable.
+///
+/// # Example
+///
+/// Note that the quotes are part of the output string.
+///
+/// ```slt
+/// query T
+/// select quote_literal(E'O\'Reilly')
+/// ----
+/// 'O''Reilly'
+///
+/// query T
+/// select quote_literal(E'C:\\Windows\\')
+/// ----
+/// E'C:\\Windows\\'
+/// ```
+#[function("quote_literal(varchar) -> varchar")]
+pub fn quote_literal(s: &str, writer: &mut impl Write) {
+    if s.contains('\\') {
+        // use escape format: E'...'
+        write!(writer, "E").unwrap();
+    }
+    write!(writer, "'").unwrap();
+    for c in s.chars() {
+        match c {
+            '\'' => write!(writer, "''").unwrap(),
+            '\\' => write!(writer, "\\\\").unwrap(),
+            _ => write!(writer, "{}", c).unwrap(),
+        }
+    }
+    write!(writer, "'").unwrap();
+}
+
+/// `quote_literal(value anyelement)`
+///
+/// Coerce the given value to text and then quote it as a literal.
+/// Embedded single-quotes and backslashes are properly doubled.
+///
+/// # Example
+///
+/// ```slt
+/// select quote_literal(42.5)
+/// ----
+/// 42.5
+///
+/// select quote_literal('hello'::bytea);
+/// --------------
+/// E'\\x68656c6c6f'
+///
+/// select quote_literal('{"hello":"world","foo":233}'::jsonb);
+/// ----
+/// '{"foo": 233, "hello": "world"}'
+/// ```
+#[function("quote_literal(any) -> varchar")]
+pub fn quote_literal_any(v: ScalarRefImpl<'_>, writer: &mut impl Write) -> Result<()> {
```

Comment:
The current implementation is not PostgreSQL-compatible.

Why not follow PostgreSQL and just rewrite to cast-text in frontend?
https://github.com/postgres/postgres/blob/REL_16_1/src/include/catalog/pg_proc.dat#L4641

```
test=# select quote_literal(42.5);
 quote_literal 
---------------
 '42.5'
(1 row)
```

Also note how this `any` variant is different from the `text` variant: The latter is `immutable` while the former is `stable`, meaning it can return different results given same input when session variables change (eg timezone, bytea style). As of now we classify both as `pure` so there is no effective difference, but it is better to code less and achieve better compatibility.

Reply:
Thanks for pointing out! I will enable these regression tests and add more to cover numeric and datetime types, etc.